File: Two Sum.js
/*
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

Complex O(n)
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    var hash = {};
    var complement = 0;
    var element = 0;

    for(var i = 0; i < nums.length; i++){
	  //Store the element in the array for quick access
	  element = nums[i];
	  //We need to check the complement for the target
	  complement = target - element;
	  //If we saw the complement, we have found the indexs
      if(complement in hash){
      	return [hash[complement], i];
      }else{ //We store the element with the index it has
      	hash[element] = i;
      }
    }
};

nums = [2, 7, 11, 15];
target = 9;
console.log(twoSum(nums,target));

File: Java/twoSum.java
/*
Given an array of integers nums and an integer target, return indices of the two numbers
such that they add up to target.

You may assume that each input would have exactly one solution, and you may not
use the same element twice.

You can return the answer in any order.

Example 1:
Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Output: Because nums[0] + nums[1] == 9, we return [0, 1].

Example 2:
Input: nums = [3,2,4], target = 6
Output: [1,2]

Example 3:
Input: nums = [3,3], target = 6
Output: [0,1]

https://leetcode.com/problems/two-sum/
*/

// Time O(N)
// Space O(N)
import java.util.HashMap; // import HashMap class
import java.util.Map; // import Map Interface
import java.util.ArrayList;
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int complement = target - num;
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };   
            } else {
                map.put(num, i);
            }
        }
        return new int[]{};
    }
}
